src/librustc/ty/maps/on_disk_cache.rs

```diff
@@ -249,6 +322,142 @@ impl<'a> SpecializedDecoder<Span> for CacheDecoder<'a> {
+// sessions, to map the old DefId to the new one.
+impl<'a, 'tcx, 'x> SpecializedDecoder<DefId> for CacheDecoder<'a, 'tcx, 'x> {
+    fn specialized_decode(&mut self) -> Result<DefId, Self::Error> {
+        // Decode the unmapped CrateNum
+        let prev_cnum = CrateNum::default_decode(self)?;
+
+        // Decode the unmapped DefIndex
+        let def_index = DefIndex::default_decode(self)?;
+
+        // Unmapped CrateNum and DefIndex are valid keys for the *cached*
+        // DefPathTables, so we use them to look up the DefPathHash.
+        let def_path_hash = self.prev_def_path_tables[prev_cnum.index()]
+                                .def_path_hash(def_index);
+
+        // Using the DefPathHash, we can lookup the new DefId
+        Ok(self.tcx().def_path_hash_to_def_id.as_ref().unwrap()[&def_path_hash])
+    }
+}
+
+impl<'a, 'tcx, 'x> SpecializedDecoder<hir::HirId> for CacheDecoder<'a, 'tcx, 'x> {
+    fn specialized_decode(&mut self) -> Result<hir::HirId, Self::Error> {
+        // Decode the unmapped DefIndex of the HirId.
+        let def_index = DefIndex::default_decode(self)?;
+
+        // Use the unmapped DefIndex to look up the DefPathHash in the cached
+        // DefPathTable. For HirIds we know that we always have to look in the
+        // *local* DefPathTable.
+        let def_path_hash = self.prev_def_path_tables[LOCAL_CRATE.index()]
+                                .def_path_hash(def_index);
+
+        // Use the DefPathHash to map to the current DefId.
+        let def_id = self.tcx()
+                         .def_path_hash_to_def_id
+                         .as_ref()
+                         .unwrap()[&def_path_hash];
+
+        // The ItemLocalId needs no remapping.
+        let local_id = hir::ItemLocalId::decode(self)?;
+
+        // Reconstruct the HirId and look up the corresponding NodeId in the
+        // context of the current session.
+        Ok(hir::HirId {
+            owner: def_id.index,
+            local_id
+        })
+    }
+}
+
+// NodeIds are not stable across compilation sessions, so we store them in their
+// HirId representation. This allows use to map them to the current NodeId.
+impl<'a, 'tcx, 'x> SpecializedDecoder<NodeId> for CacheDecoder<'a, 'tcx, 'x> {
+    fn specialized_decode(&mut self) -> Result<NodeId, Self::Error> {
+        let hir_id = hir::HirId::decode(self)?;
+        Ok(self.tcx().hir.hir_to_node_id(hir_id))
+    }
+}
+
+impl<'a, 'tcx, 'x> SpecializedDecoder<ty::Ty<'tcx>> for CacheDecoder<'a, 'tcx, 'x> {
+    fn specialized_decode(&mut self) -> Result<ty::Ty<'tcx>, Self::Error> {
+        ty_codec::decode_ty(self)
+    }
+}
+
+impl<'a, 'tcx, 'x> SpecializedDecoder<ty::GenericPredicates<'tcx>>
+for CacheDecoder<'a, 'tcx, 'x> {
+    fn specialized_decode(&mut self) -> Result<ty::GenericPredicates<'tcx>, Self::Error> {
+        ty_codec::decode_predicates(self)
+    }
+}
+
+impl<'a, 'tcx, 'x> SpecializedDecoder<&'tcx Substs<'tcx>> for CacheDecoder<'a, 'tcx, 'x> {
```

Comment:
If we forget one of these impls, what happens? I think it panics? (These impls *are* copy-n-paste, I assume.)

Reply:
Yeah, most of them should panic.